### PR TITLE
fix (parsers/json): do not escape double quotes if already escaped

### PR DESCRIPTION
**Pull request by dmenini**
This PR fixes an issues I found when upgrading to a more recent version of Langchain. I was using 0.0.142 before, and this issue popped up already when the `_custom_parser` was added to `output_parsers/json`.

Anyway, the issue is that the parser tries to escape quotes when they are double-escaped (e.g. `\\"`), leading to OutputParserException. 
This is particularly undesired in my app, because I have an Agent that uses a single input Tool, which expects as input a JSON string with the structure: 
```python
{
    "foo": string,
    "bar": string
}
```
The LLM (GPT3.5) response is (almost) always something like `"action_input": "{\\"foo\\": \\"bar\\", \\"bar\\": \\"foo\\"}"` and since the upgrade this is not correctly parsed.

### CODE REVIEW

File: libs/langchain/tests/unit_tests/output_parsers/test_json.py [hunk 1]
```diff
@@ -131,3 +138,11 @@ def test_parse_json_with_code_blocks() -> None:
         "action": "Final Answer",
         "action_input": '```bar\n<div id="1" class="value">\n\ttext\n</div>```',
     }
+
+
+def test_parse_json_with_escaped_json() -> None:
+    parsed = parse_json_markdown(JSON_WITH_DOUBLED_ESCAPED_JSON_QUOTES)
+    assert parsed == {
+        "action": "Final Answer",
+        "action_input": '{"foo": "bar", "bar": "foo"}'
```
Comment (baskaryan) [inline libs/langchain/tests/unit_tests/output_parsers/test_json.py:147]: do we not want this to be parsed to a dict?

Comment (dmenini) [reply libs/langchain/tests/unit_tests/output_parsers/test_json.py:147]: So far I noticed that:
- `"action_input": "{\\"foo\\": \\"bar\\", \\"bar\\": \\"foo\\"}"` -> raises OutputParserException
- `"action_input": "{\"foo\": \"bar\", \"bar\": \"foo\"}"` -> `"action_input": "{"foo": "bar", "bar": "foo"}"` (ok)
- `"action_input": "{"foo": "bar", "bar": "foo"}"` -> `"action_input": "{"foo": "bar", "bar": "foo"}"` (ok)
- `"action_input": {"foo": "bar", "bar": "foo"}` -> `"action_input": {"foo": "bar", "bar": "foo"}` (ok)

As you can see, the first 3 cases return a string, while the latter returns a dict. What I did in this PR was simply to fix the first case. 
I believe that the choice of which pattern to use must be left to the final user: since the LLM can generate any of these slightly different inputs despite a very specific prompt, IMO it's better to cover more edge cases and let the user decide where/how to read the args inside the tool.

Btw I will add a few more tests for the aforementioned cases.

Comment (baskaryan) [reply libs/langchain/tests/unit_tests/output_parsers/test_json.py]: ah yea i misread the test case, all the above make sense to me! and thanks for the extra test cases!
